In [2]:
import pandas as pd

df = pd.read_csv('Urban Air Quality and Health Impact Dataset.csv')
df

datetime  datetimeEpoch     tempmax    tempmin       temp  \
0    2024-09-07   1.725692e+09  106.100000  91.000000  98.500000   
1    2024-09-08   1.725779e+09  103.900000  87.000000  95.400000   
2    2024-09-09   1.725865e+09  105.000000  83.900000  94.700000   
3    2024-09-10   1.725952e+09  106.100000  81.200000  93.900000   
4    2024-09-11   1.726038e+09  106.100000  82.100000  94.000000   
..          ...            ...         ...        ...        ...   
995  2024-09-18   1.726633e+09   76.060546  64.359387  69.002142   
996  2024-09-17   1.726550e+09   68.409198  65.939319  66.567410   
997  2024-09-12   1.726122e+09   69.756690  65.286919  65.919492   
998  2024-09-14   1.726284e+09   77.106797  61.481724  68.106569   
999  2024-09-18   1.726618e+09   90.923080  79.296868  81.636991   

     feelslikemax  feelslikemin  feelslike        dew   humidity  ...  \
0      104.000000     88.100000  95.900000  51.500000  21.000000  ...   
1      100.500000     84.700000  92.300000  48.700000  21.500000  ...   
2       99.900000     81.600000  90.600000  41.700000  16.900000  ...   
3      100.600000     79.500000  89.800000  39.100000  15.700000  ...   
4      101.000000     80.000000  90.000000  40.100000  15.900000  ...   
..            ...           ...        ...        ...        ...  ...   
995     77.673823     63.510920  67.003338  59.936776  73.189130  ...   
996     68.956722     64.805635  65.992526  59.010257  74.137401  ...   
997     68.158536     63.662942  67.313322  62.024442  84.650482  ...   
998     76.426959     60.901526  68.094309  63.169608  86.860261  ...   
999     94.180423     78.071851  84.987113  73.393045  74.734715  ...   

             City  Temp_Range  Heat_Index Severity_Score  Condition_Code  \
0         Phoenix   15.100000   95.918703       4.430000             NaN   
1         Phoenix   16.900000   92.281316       3.880000             0.0   
2         Phoenix   21.100000   90.599165       3.630000             0.0   
3         Phoenix   24.900000   89.638811       2.851200             0.0   
4         Phoenix   24.000000   89.760414       3.390800             0.0   
..            ...         ...         ...            ...             ...   
995  Philadelphia   12.886665   71.837558       1.957318             NaN   
996     San Diego    2.613629   72.463491       2.537413             0.0   
997     San Diego    4.598936   67.560060       3.595470             NaN   
998   Los Angeles   15.477717   67.930437       3.498942             0.0   
999       Houston   11.017871   86.802712       3.040020             0.0   

     Month  Season  Day_of_Week  Is_Weekend  Health_Risk_Score  
0      9.0    Fall     Saturday        True          10.522170  
1      9.0    Fall       Sunday        True          10.062332  
2      9.0    Fall       Monday       False           9.673387  
3      9.0    Fall      Tuesday       False           9.411519  
4      9.0    Fall    Wednesday       False           9.515179  
..     ...     ...          ...         ...                ...  
995    9.0    Fall    Wednesday       False           8.750142  
996    9.0    Fall      Tuesday       False           9.118198  
997    9.0    Fall     Thursday       False           9.880093  
998    9.0    Fall     Saturday        True           9.561602  
999    9.0    Fall    Wednesday       False          10.978044  

[1000 rows x 46 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 46 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   datetime           1000 non-null   object 
 1   datetimeEpoch      1000 non-null   float64
 2   tempmax            1000 non-null   float64
 3   tempmin            1000 non-null   float64
 4   temp               1000 non-null   float64
 5   feelslikemax       1000 non-null   float64
 6   feelslikemin       1000 non-null   float64
 7   feelslike          1000 non-null   float64
 8   dew                1000 non-null   float64
 9   humidity           1000 non-null   float64
 10  precip             1000 non-null   float64
 11  precipprob         1000 non-null   float64
 12  precipcover        1000 non-null   float64
 13  preciptype         378 non-null    object 
 14  snow               1000 non-null   float64
 15  snowdepth          929 non-null    float64
 16  windgust           1000 n

In [4]:
df.head()


datetime  datetimeEpoch  tempmax  tempmin  temp  feelslikemax  \
0  2024-09-07   1.725692e+09    106.1     91.0  98.5         104.0   
1  2024-09-08   1.725779e+09    103.9     87.0  95.4         100.5   
2  2024-09-09   1.725865e+09    105.0     83.9  94.7          99.9   
3  2024-09-10   1.725952e+09    106.1     81.2  93.9         100.6   
4  2024-09-11   1.726038e+09    106.1     82.1  94.0         101.0   

   feelslikemin  feelslike   dew  humidity  ...     City  Temp_Range  \
0          88.1       95.9  51.5      21.0  ...  Phoenix        15.1   
1          84.7       92.3  48.7      21.5  ...  Phoenix        16.9   
2          81.6       90.6  41.7      16.9  ...  Phoenix        21.1   
3          79.5       89.8  39.1      15.7  ...  Phoenix        24.9   
4          80.0       90.0  40.1      15.9  ...  Phoenix        24.0   

   Heat_Index Severity_Score  Condition_Code  Month  Season  Day_of_Week  \
0   95.918703         4.4300             NaN    9.0    Fall     Saturday   
1   92.281316         3.8800             0.0    9.0    Fall       Sunday   
2   90.599165         3.6300             0.0    9.0    Fall       Monday   
3   89.638811         2.8512             0.0    9.0    Fall      Tuesday   
4   89.760414         3.3908             0.0    9.0    Fall    Wednesday   

   Is_Weekend  Health_Risk_Score  
0        True          10.522170  
1        True          10.062332  
2       False           9.673387  
3       False           9.411519  
4       False           9.515179  

[5 rows x 46 columns]

In [5]:
df.corr(numeric_only=True)

datetimeEpoch   tempmax   tempmin      temp  feelslikemax  \
datetimeEpoch           1.000000 -0.009753  0.152861  0.045901     -0.015768   
tempmax                -0.009753  1.000000  0.808405  0.936353      0.952111   
tempmin                 0.152861  0.808405  1.000000  0.944321      0.820330   
temp                    0.045901  0.936353  0.944321  1.000000      0.920431   
feelslikemax           -0.015768  0.952111  0.820330  0.920431      1.000000   
feelslikemin            0.164148  0.801754  0.988296  0.939750      0.821445   
feelslike               0.055531  0.912584  0.939621  0.976463      0.942921   
dew                     0.046728  0.045593  0.258251  0.150877      0.262477   
humidity                0.000268 -0.609694 -0.396124 -0.538453     -0.418409   
precip                  0.053142 -0.047861  0.129018  0.041237      0.038671   
precipprob              0.116485  0.007802  0.260121  0.149406      0.137801   
precipcover            -0.077928  0.010920  0.199713  0.127998      0.167734   
snow                         NaN       NaN       NaN       NaN           NaN   
snowdepth                    NaN       NaN       NaN       NaN           NaN   
windgust               -0.374833  0.231635  0.251239  0.275448      0.230291   
windspeed              -0.202433 -0.039331  0.090775  0.035191     -0.045441   
winddir                -0.096606 -0.259833 -0.308712 -0.302600     -0.319493   
pressure                0.029477 -0.409537 -0.441358 -0.433163     -0.414626   
cloudcover              0.126440 -0.098614  0.167980  0.042230     -0.005445   
visibility              0.776977 -0.036896  0.079811  0.000181     -0.019039   
solarradiation         -0.274901  0.220803 -0.032257  0.095926      0.176928   
solarenergy            -0.271082  0.224988 -0.034113  0.096308      0.179844   
uvindex                -0.093891  0.163928  0.020639  0.081271      0.104218   
severerisk              0.285721  0.305707  0.366885  0.343067      0.408079   
sunriseEpoch            0.992032 -0.016824  0.146502  0.039187     -0.020071   
sunsetEpoch             0.992068 -0.008860  0.149014  0.044418     -0.015194   
moonphase               0.990523 -0.010851  0.147357  0.042298     -0.017145   
Temp_Range             -0.243548  0.465754 -0.127163  0.163660      0.377892   
Heat_Index              0.083644  0.780744  0.738725  0.800903      0.815434   
Severity_Score         -0.369008  0.277908  0.226903  0.275709      0.247352   
Condition_Code               NaN       NaN       NaN       NaN           NaN   
Month                        NaN       NaN       NaN       NaN           NaN   
Is_Weekend             -0.176575 -0.037857 -0.030096 -0.030614     -0.026096   
Health_Risk_Score      -0.111190  0.354713  0.434786  0.412288      0.518574   

                   feelslikemin  feelslike       dew  humidity    precip  ...  \
datetimeEpoch          0.164148   0.055531  0.046728  0.000268  0.053142  ...   
tempmax                0.801754   0.912584  0.045593 -0.609694 -0.047861  ...   
tempmin                0.988296   0.939621  0.258251 -0.396124  0.129018  ...   
temp                   0.939750   0.976463  0.150877 -0.538453  0.041237  ...   
feelslikemax           0.821445   0.942921  0.262477 -0.418409  0.038671  ...   
feelslikemin           1.000000   0.942326  0.290646 -0.375796  0.126306  ...   
feelslike              0.942326   1.000000  0.288758 -0.425092  0.081487  ...   
dew                    0.290646   0.288758  1.000000  0.706298  0.338506  ...   
humidity              -0.375796  -0.425092  0.706298  1.000000  0.296729  ...   
precip                 0.126306   0.081487  0.338506  0.296729  1.000000  ...   
precipprob             0.266224   0.213900  0.485130  0.346617  0.610299  ...   
precipcover            0.198450   0.198866  0.468846  0.371158  0.652622  ...   
snow                        NaN        NaN       NaN       NaN       NaN  ...   
snowdepth                   NaN        NaN       NaN       NaN       NaN  ...   
win

In [6]:
correlation_matrix = df[['tempmax', 'humidity', 'Health_Risk_Score']].corr()
print(correlation_matrix)


                    tempmax  humidity  Health_Risk_Score
tempmax            1.000000 -0.609694           0.354713
humidity          -0.609694  1.000000           0.186211
Health_Risk_Score  0.354713  0.186211           1.000000


In [7]:
import matplotlib.pyplot as plt
from scipy import stats

plt.figure(figsize=(15, 5))

plt.subplot(1, 3, 1)
plt.scatter(df['tempmax'], df['humidity'], alpha=0.7)
plt.xlabel('Max Temperature (°F)')
plt.ylabel('Humidity (%)')

m1, b1, r1, p1, err1 = stats.linregress(df['tempmax'], df['humidity'])
plt.plot(df['tempmax'], m1 * df['tempmax'] + b1, 'r--')
plt.title(f'Max Temp vs Humidity')

plt.subplot(1, 3, 2)
plt.scatter(df['tempmax'], df['Health_Risk_Score'], alpha=0.7)
plt.xlabel('Max Temperature (°F)')
plt.ylabel('Health Risk Score')
m2, b2, r2, p2, err2 = stats.linregress(df['tempmax'], df['Health_Risk_Score'])
plt.plot(df['tempmax'], m2 * df['tempmax'] + b2, 'r--')
plt.title(f'Max Temp vs Health Risk Score')

plt.subplot(1, 3, 3)
plt.scatter(df['humidity'], df['Health_Risk_Score'], alpha=0.7)
plt.xlabel('Humidity (%)')
plt.ylabel('Health Risk Score')
m3, b3, r3, p3, err3 = stats.linregress(df['humidity'], df['Health_Risk_Score'])
plt.plot(df['humidity'], m3 * df['humidity'] + b3, 'r--')
plt.title(f'Humidity vs Health Risk Score')

plt.tight_layout()

plt.show()


KeyboardInterrupt: 

In [27]:
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.diagnostic import het_white
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

model = ols(formula='tempmax~humidity', data=df).fit()
model.params


#TEMP AND HUMIDITY

Intercept    104.840873
humidity      -0.347536
dtype: float64

In [28]:
white_test = het_white(model.resid,  model.model.exog)
bp_test = het_breuschpagan(model.resid, model.model.exog)
output_df = pd.DataFrame(columns=['LM stat', 'LM p-value', 'F-stat', 'F p-value'])
output_df.loc['White'] = white_test
output_df.loc['Breusch-Pagan'] = bp_test
output_df.round(5)

#PAGAN/WHITE TEST

LM stat  LM p-value   F-stat  F p-value
White          19.63029     0.00005  9.98164    0.00005
Breusch-Pagan   5.08153     0.02418  5.09727    0.02418

In [29]:
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.diagnostic import het_white
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
model = ols(formula='tempmax~Health_Risk_Score', data=df).fit()
model.params

#Temp and Health Risk

Intercept            36.751614
Health_Risk_Score     4.970175
dtype: float64

In [30]:

white_test = het_white(model.resid,  model.model.exog)
bp_test = het_breuschpagan(model.resid, model.model.exog)

output_df = pd.DataFrame(columns=['LM stat', 'LM p-value', 'F-stat', 'F p-value'])
output_df.loc['White'] = white_test
output_df.loc['Breusch-Pagan'] = bp_test
output_df.round(5)

#PAGAN/WHITE TEST

LM stat  LM p-value   F-stat  F p-value
White          13.58790     0.00112  6.86688    0.00109
Breusch-Pagan   0.26099     0.60944  0.26054    0.60987

In [31]:
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.diagnostic import het_white
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
model = ols(formula='humidity~Health_Risk_Score', data=df).fit()
model.params

#Humidity and Health Risk

Intercept            12.248930
Health_Risk_Score     4.577334
dtype: float64

In [32]:

white_test = het_white(model.resid,  model.model.exog)
bp_test = het_breuschpagan(model.resid, model.model.exog)
output_df = pd.DataFrame(columns=['LM stat', 'LM p-value', 'F-stat', 'F p-value'])
output_df.loc['White'] = white_test
output_df.loc['Breusch-Pagan'] = bp_test
output_df.round(5)

#PAGAN/WHITE TEST

LM stat  LM p-value   F-stat  F p-value
White          21.17425     0.00003  10.7837    0.00002
Breusch-Pagan  10.02690     0.00154  10.1082    0.00152

In [33]:
from scipy.stats import pearsonr
corr_temp_humidity, p_temp_humidity = pearsonr(df['tempmax'], df['humidity'])
corr_temp_health, p_temp_health = pearsonr(df['tempmax'], df['Health_Risk_Score'])
corr_humidity_health, p_humidity_health = pearsonr(df['humidity'], df['Health_Risk_Score'])

print(f'Correlation between tempmax and humidity: {corr_temp_humidity}, | p-value: {p_temp_humidity}')
print(f'Correlation between tempmax and Health Risk Score: {corr_temp_health}, | p-value: {p_temp_health}')
print(f'Correlation between humidity and Health Risk Score: {corr_humidity_health}, | p-value: {p_humidity_health}')


Correlation between tempmax and humidity: -0.6096944603827863, | p-value: 7.804714289049784e-103
Correlation between tempmax and Health Risk Score: 0.35471305791273156, | p-value: 5.10409494842458e-31
Correlation between humidity and Health Risk Score: 0.18621135470289596, | p-value: 2.972281424308098e-09


**Answering questions**

Which variables seem to be correlated to each other? Do p-values help clarify this? How?

1)

The max temperature has a strong negative correlation with humidity, measuring approximately -0.61. This means that as max temperature increases, humidity tends to decrease. The p-value for this correlation is very low at 7.8e-103, suggesting this relationship is statistically significant.

The Health Risk Score shows a moderate positive correlation with max temperature at 0.35 and a weaker positive correlation with humidity at 0.19. The p-value for the correlation between max temperature and Health Risk Score is also very low at 5.1e-31, while the p-value for humidity and Health Risk Score is very low at 2.97e-9. These results indicate that higher temperatures are associated with increased health risks. The very low p-values for all these correlations—7.8e-103, 5.1e-31, and 2.97e-9, indicate that these relationships are statistically significant. This means it's highly unlikely that these correlations occurred by chance, providing strong evidence of actual relationships between the variables. The p-values clarify that the observed correlations are meaningful and not due to random variability in the data.

Some other notable correlations:

Temperature and Feels like Temperature (0.976) (shocker!)
Humidity and Dew Point (0.706) - This would especially make sense since how the more moisture in the air can lead to a more humid environment.
Health Risk Score and Heat Index (0.610) - Shows how the heat index can affect health in a more direct manner than many of the other statistics.


NOTE: I would have gone into greater detail about other correlations; however, I made the mistake of chosing a dataset with over 30+ numeric variables. 

Which assumptions are met in each correlation? Not met? Give some of your explanation, and use p-values when needed.


2) 
**Temperature and Humidity:**
For the correlation between temperature (tempmax) and humidity, the correlation coefficient is approximately -0.61, indicating a strong negative correlation. The p-value is very low, meaning we can confidently say that the two are statistically significant. This suggests that as temperature increases, humidity tends to decrease. In terms of assumptions, linearity is present, as the scatter plot shows a downward trend. However, equal variance did not pass the assumption due to the Pagan and White test p-values being so low. This shows varying spread at different temperature levels.

**Temperature and Health Risk Score:**
The correlation between temperature (tempmax) and Health Risk Score shows a slightly positive correlation of about 0.35, with a very low p-value, indicating a statistically significant relationship. This suggests that higher temperatures are associated with higher health risk scores; however, it doesn't tell the full story. The assumptions for this correlation appear to be mostly met, as there seems to be a linear relationship but there tends to be a fair amount of outliers. While the data suggests a trend, the potential there is a lot up for debate. The White test indicates significant heteroscedasticity with a very low p-value, suggesting that the spread of the residuals is not consistent throughout the dataset. The Breusch-Pagan test exhibits a high p-value, which indicates that the linear relationship shows equal variance in the residuals, suggesting that in a linear sense, there is a equal variance. This creates confusion because, while there appears to be a linear relationship, the assumption of homoscedasticity—where the residuals have a constant variance—does not hold true.

**Humidity and Health Risk Score:**
For the relationship between humidity and Health Risk Score, the correlation coefficient is approximately 0.19, with a p-value that is also very low. This indicates a weak positive correlation, suggesting that as humidity increases, health risk scores tend to rise slightly. However, this relationship is significantly less strong than the others, and while the linearity assumption is likely met, it’s important to note that the weak correlation indicates that humidity may not be a strong predictor of health risks. Additionally, the assumption of homoscedasticity should be considered, as the variance of the residuals might not be constant across different levels of humidity.

3)


The results from the analysis are pretty surprising but also kind of what I expected in environmental health studies. The strong negative correlation between temperature and humidity makes sense, since higher temperatures usually mean lower humidity levels in many places. But the moderately positive correlation between temperature and Health Risk Score is interesting; it shows that when temperatures go up, health risks also go up, probably because of heat-related illnesses.

The weak positive correlation between humidity and Health Risk Score is less surprising. Humidity can definitely make things uncomfortable and can cause breathing issues, but it’s not as directly related to health risks as temperature is. Honestly, I thought the correlation would be higher because of how sticky and uncomfortable humidity can feel. Overall, while some results match what you usually see in environmental health, the weak link with humidity was unexpected for me.

NameError: name 'df' is not defined